### PR TITLE
rpc: Prevent unloading a wallet when rescanning

### LINKED ISSUE #26463: [wallet] when unloading a wallet which still rescans the wallet, the command should fail but bitcoind gets in a buggy state
I am loading descriptors into bitcoind using the rpcclient with the function: `importdescriptors`, it automatically starts rescanning the blockchain. Now when I try to unload the wallet, bitcoin-cli will allow me to do it (wallet disappears from the `listwallet` section, but it will not abort the rescan. Thats a problem bc now when trying to reload the wallet, the wallet sqlite db is locked by bitcoind and this wallet is unusable unitl the rescan is over. `abortrescan` will not work bc the wallet is not loaded anymore according to `bitcoin-cli listwallets`. The only solution for me was to delete the wallet dir and create a new wallet with the same name. Now everything works as expected, but as soon as I want to unload the wallet again, bitcoind crashes

The `unloadwallet` function should internally abort the rescan and unload the wallet, so that it can be reloaded afterwards without waiting for the rescan. Or it should fail and leading the user to abort the rescan with `abortrescan`

* Create a descriptor wallet
* Import a descriptor and unload the walllet while bitcoind is still rescanning
* Now you are stuck bc abortrescan will not work bc wallet is not loaded and the only solution is to delete the wallet dir
* Now Create a new wallet with the same name and import the descriptors again. When the wallet is faster synced then the old process still syncing, and you unload the wallet, bitcoind crashes

` Linux 5.10.103-v8+ #1529 SMP PREEMPT Tue Mar 8 12:26:46 GMT 2022 aarch64 GNU/Linux`

bitcoind v.23.0 (release binary for raspberry pi)

### PR DESCRIPTION
**Pull request by aureleoules**
Fixes #26463.

This PR prevents a user from unloading a wallet if it is currently rescanning.

To test:

```bash
./src/bitcoin-cli -testnet -named createwallet wallet_name=wo disable_private_keys=true
./src/bitcoin-cli -testnet -rpcwallet=wo importdescriptors '[{
	  "desc": "addr(mmcuW74MyJUZuLnWXGQLoNXPrS9RbFz6gD)#tpnrahgc",
      "timestamp": 0,
      "active": false,
      "internal": false,
      "next": 0
}]'
./src/bitcoin-cli -testnet unloadwallet wo
error code: -4
error message:
Wallet is currently rescanning. Abort existing rescan or wait.

### CODE REVIEW

**Review by stickies-v — COMMENTED** [review]
Current behaviour seems to be that `unloadwallet` waits until the scan is finished. Changing that by throwing could affect downstream users that are expecting `unloadwallet` to just wait. I think the behaviour in this PR is better if it were implemented like that from the start. I'm not sure if it's worth changing though? ~0 on this for now.

**Review by ishaanam — COMMENTED** [review]
> Current behaviour seems to be that unloadwallet waits until the scan is finished.

This is what I have observed as well. Even so, I think that it would be better if instead of waiting for the rescan to finish (which could take hours), `CWallet::AbortRescan` is called from `unloadwallet`. This seems better because `unloadwallet` should ideally be able to unload the wallet relatively quickly regardless of the processes that the wallet is running. A warning could be given when a rescan has been aborted by `unloadwallet` so that a user is not surprised.

**Comment by aureleoules** [comment]
> Changing that by throwing could affect downstream users that are expecting unloadwallet to just wait.

Do you know any software that would use this behavior? I understand this could affect users but this seems a bit hacky.

> A warning could be given when a rescan has been aborted by unloadwallet so that a user is not surprised.

I don't really have an opinion on this. But an argument for keeping the warning in 8198c76a3008094fefbbae1b11e5d5673a993158 is that an user may not realize that the wallet is still rescanning and wouldn't want to abort it when trying to unload.
As a compromise we could add a 'force' argument to unloadwallet to abort the rescan if there is one.

**Comment by w0xlt** [comment]
> As a compromise we could add a 'force' argument to unloadwallet to abort the rescan if there is one.

I think this is a good approach.

**Comment by luke-jr** [comment]
Agree with a force option (named only / options Object).

But also, we probably shouldn't return from `unloadwallet` until the wallet is actually unloaded fully, independently of whether that's due to a rescan or otherwise.

I guess this PR is strictly better than the current behaviour, though, so Concept ACK.

**Comment by aureleoules** [comment]
I added the named `force` option that will abort the rescan in progress.
https://github.com/bitcoin/bitcoin/compare/8198c76a3008094fefbbae1b11e5d5673a993158..bee026e801cf692707477ca51505d708ef051cae

**Review by promag — COMMENTED** [review]
Tiny concept NACK.

I think `unloadwallet` should just fail if a scan is in progress. We already have that behavior/warning elsewhere: `Wallet is currently rescanning. Abort existing rescan or wait.` I understand it can be seen as a breaking change, but makes more sense than what happened in #26463.

---
Regardless of the above, the current code is still not ideal, for instance, a scan can be initiated after `IsScanning` check if `importdescriptors` is called concurrently. I think you should use `WalletRescanReserver` because it allows checking if a rescan is in progress and at the same time prevent concurrent scans.

A test case would be nice.

File: src/wallet/rpc/wallet.cpp [hunk 1]
```diff
@@ -412,6 +412,7 @@ static RPCHelpMan unloadwallet()
                 {
                     {"wallet_name", RPCArg::Type::STR, RPCArg::DefaultHint{"the wallet name from the RPC endpoint"}, "The name of the wallet to unload. If provided both here and in the RPC endpoint, the two must be identical."},
                     {"load_on_startup", RPCArg::Type::BOOL, RPCArg::Optional::OMITTED_NAMED_ARG, "Save wallet name to persistent settings and load on startup. True to add wallet to startup list, false to remove, null to leave unchanged."},
+                    {"force", RPCArg::Type::BOOL, RPCArg::Optional::OMITTED_NAMED_ARG, "Force unload of wallet. This will abort any rescan in progress."},
```
Comment (luke-jr) [inline src/wallet/rpc/wallet.cpp:415]: This is a positional parameter. See `simulaterawtransaction` for how to do named-only parameters in an options Object.

(Maybe move "load_on_startup" there too)

**Review by luke-jr — CHANGES_REQUESTED** [review]

**Comment by ziggie1984** [comment]
Thanks for fixing this issue. I like the approach more where we just fail the `unloadwallet` call but the force flag is also a good way. 

>  a scan can be initiated after IsScanning check if importdescriptors is called concurrently.

Definitly important to prevent bitcoind from crashing. I was able to bring bitcoind down when unloading the wallet (blocking in the background) deleting the walletdir and then importing the descriptors with the same wallet name again (an then there was a dependence that one sync finished before the other). So would be important if we would not allow loading a wallet which is currently in rescan (at least not with the same name).

File: src/wallet/rpc/wallet.cpp [hunk 2]
```diff
@@ -437,6 +438,14 @@ static RPCHelpMan unloadwallet()
         throw JSONRPCError(RPC_WALLET_NOT_FOUND, "Requested wallet does not exist or is not loaded");
     }
 
+    if (wallet->IsScanning()) {
+        if (request.params[2].isNull() || !request.params[2].get_bool()) {
+            throw JSONRPCError(RPC_WALLET_ERROR, "Cannot unload a wallet while it is rescanning. Retry with the force option or run abortrescan first.");
+        }
+
+        wallet->AbortRescan();
```
Comment (furszy) [inline src/wallet/rpc/wallet.cpp:446]: As `AbortRescan()` does not abort the scanning process right away (only sets a flag that is later read by the wallet's scanning thread), might be good to add a "wait until scan aborts" here.

So there is no race between the wallet removal and the scanning process.

File: src/wallet/rpc/wallet.cpp [hunk 3]
```diff
@@ -437,6 +437,11 @@ static RPCHelpMan unloadwallet()
         throw JSONRPCError(RPC_WALLET_NOT_FOUND, "Requested wallet does not exist or is not loaded");
     }
 
+    WalletRescanReserver reserver(*wallet);
```
Comment (promag) [inline src/wallet/rpc/wallet.cpp:440]: `reserver` destructor needs to be called before `UnloadWallet`, something like:
```cpp
{
    WalletRescanReserver reserver(*wallet);
    ....
    
    RemoveWallet()
}
UnloadWallet();

```

**Review by promag — CHANGES_REQUESTED** [review]
Concept ACK, `unloadwallet` fails if a rescan is in progress. Needs:
 - release notes
 - test, not sure how 
 - follow-up in the GUI

**Comment by aureleoules** [comment]
Thanks, @promag. I added release notes.

I think the test should be a follow-up as there are many other instances where this warning is not tested.

**Comment by promag** [comment]
> I think the test should be a follow-up as there are many other instances where this warning is not tested.

The general rule of thumb is that tests are added when there are changes.

**Comment by aureleoules** [comment]
> The general rule of thumb is that tests are added when there are changes.

I agree but the RPC error `Wallet is currently rescanning. Abort existing rescan or wait` has not been tested in other RPCs either and if a test were to be added, the other RPCs should be tested too and that may be out of scope.

```
grep -nrIi "Wallet is currently rescanning. Abort existing rescan or wait" ./src
src/wallet/rpc/backup.cpp:159:            throw JSONRPCError(RPC_WALLET_ERROR, "Wallet is currently rescanning. Abort existing rescan or wait.");
src/wallet/rpc/backup.cpp:254:        throw JSONRPCError(RPC_WALLET_ERROR, "Wallet is currently rescanning. Abort existing rescan or wait.");
src/wallet/rpc/backup.cpp:447:        throw JSONRPCError(RPC_WALLET_ERROR, "Wallet is currently rescanning. Abort existing rescan or wait.");
src/wallet/rpc/backup.cpp:516:        throw JSONRPCError(RPC_WALLET_ERROR, "Wallet is currently rescanning. Abort existing rescan or wait.");
src/wallet/rpc/backup.cpp:1351:        throw JSONRPCError(RPC_WALLET_ERROR, "Wallet is currently rescanning. Abort existing rescan or wait.");
src/wallet/rpc/backup.cpp:1658:        throw JSONRPCError(RPC_WALLET_ERROR, "Wallet is currently rescanning. Abort existing rescan or wait.");
src/wallet/rpc/transactions.cpp:872:        throw JSONRPCError(RPC_WALLET_ERROR, "Wallet is currently rescanning. Abort existing rescan or wait.");
src/wallet/rpc/wallet.cpp:444:            throw JSONRPCError(RPC_WALLET_ERROR, "Wallet is currently rescanning. Abort existing rescan or wait.");
```

```
grep -nrIi "Wallet is currently rescanning. Abort existing rescan or wait" ./test
<empty>
```

I'll look into adding a test though.

**Comment by promag** [comment]
My suggestion is to test the just introduced error, the other are indeed out of scope.

**Comment by ziggie1984** [comment]
Is it conceptionally possible to abort the rescan, if the wallet_dir is deleted ? My tests showed me that bitcoind will happily rescan further ? Or is it somehow a user-misbehaving we should not take into account ?

**Comment by promag** [comment]
@ziggie1984 user-misbehaving

**Comment by luke-jr** [comment]
Yeah, I would think deleting data files out from under the software should be expected to crash ;)

**Comment by aureleoules** [comment]
I added a test in #26700 @promag.

**Review by w0xlt — APPROVED** [review]
ACK https://github.com/bitcoin/bitcoin/pull/26618/commits/109cbb819ddd20220a255791c40261f746260f42

**Review by kouloumos — COMMENTED** [review]
ACK 109cbb819ddd20220a255791c40261f746260f42
The implementation is consistent with how the `WalletRescanReserver` is used in the rest of the related RPCs as pointed out in https://github.com/bitcoin/bitcoin/pull/26618#pullrequestreview-1209074941 .
(check the rest of the RPCs with `git grep -n "Wallet is currently rescanning. Abort existing rescan or wait"`)

I would also prefer to see a test as part of this PR, but looking at the linked draft PR implementation it seems that the relative complexity to include a test (other test cases missing, refactor of an existing test file) justifies a follow-up.

**Review by promag — COMMENTED** [review]
ACK 109cbb819ddd20220a255791c40261f746260f42

**Comment by achow101** [comment]
ACK 109cbb819ddd20220a255791c40261f746260f42

**Comment by achow101** [comment]
This was merged

**Pull request closed without merging**